### DISCUSSION TITLE
Unit Testing With Koin Isolated Context

### DISCUSSION THREAD

**Discussion by Ynnck123** [InsertKoinIO/koin#1907] [Q&A]
Hi,

when using Koin Isolated Context like the Koin documentation suggests: 

```
// Get a Context for your Koin instance
object MyIsolatedKoinContext {

    val koinApp = koinApplication {
        // declare used modules
        modules(coffeeAppModule)
    }

    val koin = koinApp.koin 
}
```

How would you overwrite the koinApp in your unit tests to switch declarations? I am currently running into the problem that I am not able to use different implementations in my unit tests because I can only fill the default Koin context from my understanding and have no option to switch out this isolated context.

Do you have any suggestions?

_Edit: I am using field injection via overwritten `getKoin()` from the `KoinComponent` function that uses my isolated context_

**Comment by ragboyjr** [comment]
Not sure when these docs were added, but they talk about this here: https://github.com/InsertKoinIO/koin/discussions?discussions_q=is%3Aopen+

Personally, I also use a similar approach to just explicitly overriding services via the MyIsolatedKoinContext.koin instance.

**Comment by tezov** [comment]
If you don't declare your module in your context like this,

```
object KoinContext {
    lateinit var koinApplication: KoinApplication

    val koin get() = koinApplication.koin
}
```

Then somewhere else : 

```
    fun startMyIsolatedKoin() = koinApplication {
            
           what ever modules you need

        }.also { KoinContext.koinApplication = it }
        
```

Then when you test :

```
    @BeforeTest
    fun setup() {
        processActionUseCase = mock()

        // minimal injection need for this test
        KoinContext.koinApplication = koinApplication {
            modules(
                module {
                    single<ProcessActionUseCase> { processActionUseCase }
                }
            )
        }
      ....

        sut = FormSendUrlActionMiddleware(
            useCaseExecutor = useCaseExecutor,
            ...
        )

```

It will work either you inject with constructor or  implement  with your KoinComponent. 


```
interface KoinComponent : org.koin.core.component.KoinComponent {
    override fun getKoin(): Koin = KoinContext.koin
}

```

PS: Works also with scope ;)


Best regards.
